PowerShell Contributing.md

```diff
@@ -0,0 +1,479 @@
+
+`Start-PSPester -Path .\test\powershell\Modules\Microsoft.PowerShell.Core\Get-Command.Tests.ps1`
+
+You will notice that it rebuilds test tool binaries each time you use Start-PSPester.
+This takes time and usually we only need to build these binaries once.
+To suppress new builds, you add the `-SkipTestToolBuild` switch:
+
+`Start-PSPester -Path .\test\powershell\Modules\Microsoft.PowerShell.Core\Get-Command.Tests.ps1 -SkipTestToolBuild`
+
+When writing and debugging your tests, remember that you can set breakpoints and step through your Pester code, this can help you find out where things are not going as they | Should -Be :-).
+
+> Tip: If you want to learn more about VSCode debugging for PowerShell, you can read more [here on the Microsoft scripting blog](https://devblogs.microsoft.com/scripting/debugging-powershell-script-in-visual-studio-code-part-1/), or watch [this video](https://youtu.be/cSbIXmlkr8o) from Tyler Leonhardt at PSconfEU 2019.
+
+For specific tips about how to use Pester within the PowerShell repo, be sure to read this document as an addendum to the Pester documentation: <https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md>
+
+### Updating your fork
+
+So you have forked one of the GitHub repos, cloned it, tried stuff, but never created a PR. And now that fork is six months behind its ancester. How to update it? I'll describe how to do this for the PowerShell repo, this works the same for the other repos.
+
+To update your GitHub fork, you will first update your local clone, and for that, you need to tell your local repository to which remote to look. 
+
+- See which remotes are configured for your local repo: `git remote -v`
+- If there is an entry with `https://github.com/PowerShell/PowerShell.git` in it, remember its name. It will usually be called 'upstream'
+- If there is no entry present with that url, add it by typing:
+`git remote add upstream https://github.com/PowerShell/PowerShell.git`
+- Make sure there is an entry 'origin' with the url of your own GitHub url: `https://github.com/PowerShell/PowerShell.git` 
+
+Make sure you have no uncommitted work on your branch. Stash or revert any uncommitted changes, and checkout to the master branch:
+
+- To stash changes, go `git stash`
+- To checkout master, do `git checkout master`
+
+Now we need to check for changes and pull them in.
+
+- Tell git to find out if there are changes in any of its remotes: `git fetch --all -p`
+- Tell git to rebase your local repo on the 'upstream' remote (replace the name 'upstream' if necessary):
+`git rebase upstream/master`
+
+Rebasing is the way to tell git to replay your work on top of the other changes that were merged to master in the mean time. So now we have pulled in all changes from the original repo into our clone locally. These changes need to be pushed to your GitHub fork.
+
+- Tell git to push local changes to origin: `git push` or `git push origin/master`.
```

Comment:
After a rebase, you will almost always need to `git push --force-with-lease`, otherwise the changes will be rejected due to the divergent histories.

Reply:
Also, if you're just pulling changes to the master branch and not your working branch, you usually won't even need to rebase; typically `git pull upstream master` is sufficient to pull the commits in, and then you can just `git push`.

Reply:
And if you've made a mess of your master branch, you can get it back to the current upstream "clean" state with `git reset --hard upstream/master` (this will completely replace your current working branch with what's there on the targeted branch, so be sure that's what you want first; there's no `undo` for that operation that I'm presently aware of).

Reply:
This made me realise most contributors will not need to cooperate on a branch and thus don't really need to rebase, as you point out, pull will do.